include/xtensor/xgenerator.hpp

```diff
@@ -0,0 +1,494 @@
+/***************************************************************************
+* Copyright (c) 2016, Johan Mabille and Sylvain Corlay                     *
+*                                                                          *
+* Distributed under the terms of the BSD 3-Clause License.                 *
+*                                                                          *
+* The full license is in the file LICENSE, distributed with this software. *
+****************************************************************************/
+
+#ifndef xgenerator_HPP
+#define xgenerator_HPP
+
+#include <cstddef>
+#include <type_traits>
+#include <utility>
+#include <tuple>
+#include <algorithm>
+
+#include "xexpression.hpp"
+#include "xiterator.hpp"
+#include "xutils.hpp"
+
+namespace xt
+{
+
+    template <class F, class R>
+    class xgenerator_stepper;
+
+    /**************
+     * xgenerator *
+     **************/
+
+    /**
+     * @class xgenerator
+     * @brief Multidimensional function operating on indices.
+     *
+     * Th xgenerator class implements a multidimensional function
+     * operating on the supplied indices.
+     *
+     * @tparam F the function type
+     * @tparam R the return type of the function
+     */
+    template <class F, class R>
+    class xgenerator : public xexpression<xgenerator<F, R>>
```

Comment:
Indeed, `xgenerator` should probably have a `class S` template parameter, and then `using shape_type=S`.